src/animation/AnimatedComponent.js

```diff
@@ -1,53 +1,58 @@
 /* @flow strict-local */
-import React, { PureComponent } from 'react';
+import React, { useRef, useCallback, useLayoutEffect } from 'react';
 import type { Node as React$Node } from 'react';
 import { Animated, Easing } from 'react-native';
 
+import { usePrevious } from '../reactUtils';
 import type { Style } from '../types';
 
 type Props = $ReadOnly<{|
   stylePropertyName: string,
   fullValue: number,
   children: React$Node,
   style?: Style,
-  visible: boolean,
-  useNativeDriver: boolean,
-  delay: number,
+  visible?: boolean,
+  useNativeDriver?: boolean,
+  delay?: number,
 |}>;
 
-export default class AnimatedComponent extends PureComponent<Props> {
-  static defaultProps = {
-    visible: true,
-    useNativeDriver: true,
-    delay: 0,
-  };
-
-  animatedValue = new Animated.Value(0);
-
-  animate() {
-    Animated.timing(this.animatedValue, {
-      toValue: this.props.visible ? this.props.fullValue : 0,
-      delay: this.props.delay,
+/**
+ * Animates the specified style property on visibility change.
+ */
+export default function AnimatedComponent(props: Props): React$Node {
+  const {
+    visible = true,
+    useNativeDriver = true,
+    delay = 0,
+    fullValue,
+    children,
+    stylePropertyName,
+    style,
+  } = props;
+
+  const prevVisible = usePrevious(visible);
+
+  const animatedValue = useRef(new Animated.Value(0));
+
+  const animate = useCallback(() => {
+    Animated.timing(animatedValue.current, {
+      toValue: visible ? fullValue : 0,
+      delay,
       duration: 300,
-      useNativeDriver: this.props.useNativeDriver,
+      useNativeDriver,
       easing: Easing.out(Easing.poly(4)),
     }).start();
-  }
+  }, [delay, fullValue, useNativeDriver, visible]);
 
-  componentDidMount() {
-    this.animate();
-  }
+  useLayoutEffect(() => {
+    if (prevVisible !== visible) {
+      animate();
```

Comment:
Yeah, that does seem better.

Looking at existing uses of this component, there's just one: it's in `ComposeMenu`, and it always starts with `visible` false. So we're currently not exercising this case at all. That's good because it means we don't have anything relying somehow on the current kind-of-quirky behavior.